File: JAVA/SchorrSignatureAlgorithm.java
import java.math.BigInteger;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;

/**
 * Triển khai thuật toán Chữ ký số Schnorr
 * 
 * Thuật toán Schnorr là một sơ đồ chữ ký số dựa trên Bài toán Logarit Rời Rạc
 * (DLP)
 * Có 3 giai đoạn chính:
 * 1. Key Generation: Phát sinh cặp khóa (khóa bí mật, khóa công khai)
 * 2. Signing: Tạo chữ ký cho một thông điệp
 * 3. Verification: Kiểm tra tính hợp lệ của chữ ký
 */
public class SchorrSignatureAlgorithm {
    private SchorrKeyPair keyPair;
    private final SecureRandom random;

    public SchorrSignatureAlgorithm() {
        this.random = new SecureRandom();
    }

    /**
     * GIAI ĐOẠN 1: Phát sinh khóa (Key Generation)
     * 
     * Quá trình:
     * 1. Nhận các tham số hệ thống (p, q, g)
     * 2. Chọn khóa bí mật x ngẫu nhiên từ [1, q-1]
     * 3. Tính khóa công khai y = g^x mod p
     * 
     * @return SchorrKeyPair chứa cặp khóa công khai/bí mật
     */
    public SchorrKeyPair generateKeys() {
        // Phát sinh các tham số hệ thống
        SchorrSignatureParams params = SchorrSignatureParams.generateSystemParameters();

        // Bước 1: Chọn khóa bí mật x ngẫu nhiên từ [1, q-1]
        BigInteger q = params.getQ();
        BigInteger x = new BigInteger(q.bitLength(), random);
        while (x.compareTo(BigInteger.ONE) < 0 || x.compareTo(q) >= 0) {
            x = new BigInteger(q.bitLength(), random);
        }

        // Bước 2: Tính khóa công khai y = g^x mod p
        BigInteger g = params.getG();
        BigInteger p = params.getP();
        BigInteger y = g.modPow(x, p);

        // Lưu cặp khóa
        this.keyPair = new SchorrKeyPair(x, y, params);
        return keyPair;
    }

    /**
     * GIAI ĐOẠN 2: Tạo chữ ký (Signing)
     * 
     * Quá trình ký với thông điệp M:
     * 1. Chọn số ngẫu nhiên k từ [1, q-1]
     * 2. Tính r = g^k mod p
     * 3. Tính e = H(M || r) sử dụng SHA-256
     * 4. Tính s = k + e*x mod q
     * 5. Trả về chữ ký (s, e)
     * 
     * @param message Thông điệp cần ký
     * @return SchorrSignature chứa cặp (s, e)
     * @throws Exception nếu lỗi xảy ra trong quá trình ký
     */
    public SchorrSignature sign(String message) throws Exception {
        if (keyPair == null) {
            throw new IllegalStateException("Chưa phát sinh khóa. Vui lòng gọi generateKeys() trước.");
        }

        SchorrSignatureParams params = keyPair.getParams();
        BigInteger x = keyPair.getPrivateKey(); // Khóa bí mật
        BigInteger p = params.getP();
        BigInteger q = params.getQ();
        BigInteger g = params.getG();

        // Bước 1: Chọn số ngẫu nhiên k từ [1, q-1]
        BigInteger k = new BigInteger(q.bitLength(), random);
        while (k.compareTo(BigInteger.ONE) < 0 || k.compareTo(q) >= 0) {
            k = new BigInteger(q.bitLength(), random);
        }

        // Bước 2: Tính r = g^k mod p
        BigInteger r = g.modPow(k, p);

        // Bước 3: Tính e = H(M || r) sử dụng SHA-256
        // Nối thông điệp M với giá trị r
        String messageWithR = message + r.toString(16);
        BigInteger e = hashSHA256(messageWithR, q);

        // Bước 4: Tính s = k + e*x mod q
        // s = (k + e*x) mod q
        BigInteger s = k.add(e.multiply(x)).mod(q);

        // Bước 5: Trả về chữ ký (s, e)
        return new SchorrSignature(s, e);
    }

    /**
     * GIAI ĐOẠN 3: Kiểm tra chữ ký (Verification)
     * 
     * Quá trình xác minh chữ ký (s, e) cho thông điệp M:
     * 1. Tính r' = g^s * y^(-e) mod p
     * Hoặc tương đương: r' = g^s * y^(q-1-e) mod p
     * 2. Tính e' = H(M || r')
     * 3. Kiểm tra nếu e' == e thì chữ ký hợp lệ
     * 
     * Lý thuyết: Nếu chữ ký hợp lệ thì:
     * g^s * y^(-e) = g^(k + e*x) * (g^x)^(-e) = g^k = r
     * 
     * @param message   Thông điệp gốc
     * @param signature Chữ ký cần xác minh
     * @return true nếu chữ ký hợp lệ, false nếu không
     * @throws Exception nếu lỗi xảy ra
     */
    public boolean verify(String message, SchorrSignature signature) throws Exception {
        if (keyPair == null) {
            throw new IllegalStateException("Chưa phát sinh khóa. Vui lòng gọi generateKeys() trước.");
        }

        SchorrSignatureParams params = keyPair.getParams();
        BigInteger y = keyPair.getPublicKey(); // Khóa công khai
        BigInteger p = params.getP();
        BigInteger q = params.getQ();
        BigInteger g = params.getG();
        BigInteger s = signature.getS();
        BigInteger e = signature.getE();

        // Bước 1: Tính r' = g^s * y^(-e) mod p
        // y^(-e) = y^(q-e) vì y^q = 1 (theo định lý Fermat)
        // Do đó: y^(-e) = y^(q-e)

        // Tính g^s mod p
        BigInteger gs = g.modPow(s, p);

        // Tính y^(-e) mod p, tức là y^(q - e) mod p
        // Điều này hoạt động vì y^q ≡ 1 (mod p) theo định lý Fermat
        BigInteger negE = e.negate().mod(q); // -e mod q
        BigInteger yInvE = y.modPow(negE, p);

        // Tính r' = g^s * y^(-e) mod p
        BigInteger rPrime = gs.multiply(yInvE).mod(p);

        // Bước 2: Tính e' = H(M || r')
        String messageWithRPrime = message + rPrime.toString(16);
        BigInteger ePrime = hashSHA256(messageWithRPrime, q);

        // Bước 3: Kiểm tra e' == e
        return ePrime.equals(e);
    }

    /**
     * Hàm băm SHA-256 để tính hash của thông điệp
     * Kết quả được chuyển đổi thành BigInteger modulo q
     * 
     * @param input Chuỗi cần băm
     * @param q     Modulo để giảm kích thước hash
     * @return BigInteger đại diện cho hash
     * @throws NoSuchAlgorithmException nếu SHA-256 không khả dụng
     */
    private BigInteger hashSHA256(String input, BigInteger q) throws NoSuchAlgorithmException {
        // Tạo đối tượng MessageDigest cho SHA-256
        MessageDigest digest = MessageDigest.getInstance("SHA-256");

        // Tính hash của input
        byte[] hashBytes = digest.digest(input.getBytes());

        // Chuyển byte[] thành BigInteger
        BigInteger hashValue = new BigInteger(1, hashBytes);

        // Lấy modulo q để đảm bảo e nằm trong [0, q-1]
        return hashValue.mod(q);
    }

    // Getters
    public SchorrKeyPair getKeyPair() {
        return keyPair;
    }

    public boolean hasKeyPair() {
        return keyPair != null;
    }
}

File: C#/SchorrSignatureAlgorithm.cs
using System;
using System.Numerics;
using System.Security.Cryptography;
using System.Text;

namespace Schnorr
{
    public class SchorrSignatureAlgorithm
    {
        private SchorrKeyPair keyPair;
        private readonly RandomNumberGenerator rng = RandomNumberGenerator.Create();

        public SchorrSignatureAlgorithm() { }

        public SchorrKeyPair GenerateKeys(int qBitLength = 256)
        {
            var parameters = SchorrSignatureParams.GenerateSystemParameters(qBitLength);
            BigInteger q = parameters.Q;

            BigInteger x = RandomInRange(BigInteger.One, q - 1);
            BigInteger y = BigInteger.ModPow(parameters.G, x, parameters.P);

            keyPair = new SchorrKeyPair(x, y, parameters);
            return keyPair;
        }

        public SchorrSignature Sign(string message)
        {
            if (keyPair == null)
                throw new InvalidOperationException(
                    "Key pair not generated. Call GenerateKeys() first."
                );
            var p = keyPair.Params.P;
            var q = keyPair.Params.Q;
            var g = keyPair.Params.G;
            var x = keyPair.PrivateKey;

            BigInteger k = RandomInRange(BigInteger.One, q - 1);
            BigInteger r = BigInteger.ModPow(g, k, p);
            BigInteger e = HashToBigInteger(message, r, q);
            BigInteger s = (k + e * x) % q;
            return new SchorrSignature(s, e);
        }

        public bool Verify(string message, SchorrSignature signature)
        {
            if (keyPair == null)
                throw new InvalidOperationException(
                    "Key pair not generated. Call GenerateKeys() first."
                );
            var p = keyPair.Params.P;
            var q = keyPair.Params.Q;
            var g = keyPair.Params.G;
            var y = keyPair.PublicKey;

            BigInteger s = signature.S;
            BigInteger e = signature.E;

            BigInteger gs = BigInteger.ModPow(g, s, p);
            // y^(-e) mod p == y^(q - e) mod p
            BigInteger negE = (q - (e % q)) % q;
            BigInteger yInvE = BigInteger.ModPow(y, negE, p);
            BigInteger rPrime = (gs * yInvE) % p;
            BigInteger ePrime = HashToBigInteger(message, rPrime, q);
            return ePrime == e;
        }

        private BigInteger HashToBigInteger(string message, BigInteger r, BigInteger q)
        {
            using var sha = SHA256.Create();
            byte[] msg = Encoding.UTF8.GetBytes(message);
            byte[] rBytes = ToBigEndian(r);
            byte[] input = new byte[msg.Length + rBytes.Length];
            Buffer.BlockCopy(msg, 0, input, 0, msg.Length);
            Buffer.BlockCopy(rBytes, 0, input, msg.Length, rBytes.Length);
            byte[] hash = sha.ComputeHash(input);
            BigInteger hv = new BigInteger(
                hash.Reverse().ToArray().Concat(new byte[] { 0 }).ToArray()
            );
            return hv % q;
        }

        private BigInteger RandomInRange(BigInteger minInclusive, BigInteger maxInclusive)
        {
            if (minInclusive > maxInclusive)
                throw new ArgumentException("min > max");
            BigInteger range = maxInclusive - minInclusive + 1;
            BigInteger r;
            do
            {
                r = RandomBigIntegerBelow(range);
            } while (r < 0 || r >= range);
            return minInclusive + r;
        }

        private BigInteger RandomBigIntegerBelow(BigInteger bound)
        {
            if (bound <= 0)
                return 0;
            int bits = BigIntegerExtensions.GetBitLengthExt(bound);
            int bytes = (bits + 7) / 8;
            byte[] b = new byte[bytes];
            BigInteger r;
            do
            {
                rng.GetBytes(b);
                // set top bit to limit size
                int topBits = bits - 8 * (bytes - 1);
                if (topBits < 8)
                {
                    int m = (1 << topBits) - 1;
                    byte mask = (byte)m;
                    b[0] &= mask;
                }
                r = new BigInteger(b.Reverse().ToArray().Concat(new byte[] { 0 }).ToArray());
            } while (r >= bound);
            return r;
        }

        private static byte[] ToBigEndian(BigInteger v)
        {
            if (v.IsZero)
                return new byte[] { 0 };
            var le = v.ToByteArray(); // little-endian two's complement
            // strip sign byte if present
            if (le[le.Length - 1] == 0)
            {
                le = le.Take(le.Length - 1).ToArray();
            }
            return le.Reverse().ToArray();
        }
    }
}
